tests/tests_pytorch/strategies/test_colossalai_strategy.py

```diff
@@ -0,0 +1,263 @@
+import os
+from typing import Any, Dict
+
+import pytest
+import torch
+import torch.nn.functional as F
+from torch import nn, Tensor
+from torch.optim import Optimizer
+from torchmetrics import Accuracy
+
+from pytorch_lightning import LightningModule, seed_everything, Trainer
+from pytorch_lightning.demos.boring_classes import BoringModel
+from pytorch_lightning.plugins.precision import ColossalAIPrecisionPlugin
+from pytorch_lightning.strategies import ColossalAIStrategy
+from pytorch_lightning.strategies.colossalai import _COLOSSALAI_AVAILABLE
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests_pytorch.helpers.datamodules import ClassifDataModule
+from tests_pytorch.helpers.runif import RunIf
+
+if _COLOSSALAI_AVAILABLE:
+    from colossalai.nn.optimizer import HybridAdam
+
+
+def test_invalid_colosalai(monkeypatch):
+    import pytorch_lightning.strategies.colossalai as colossal_strategy
+
+    monkeypatch.setattr(colossal_strategy, "_COLOSSALAI_AVAILABLE", False)
+    with pytest.raises(MisconfigurationException):
+        ColossalAIStrategy()
+
+
+@RunIf(colossalai=True)
+def test_colossalai_strategy_with_trainer_by_instance(tmpdir):
+    trainer = Trainer(fast_dev_run=True, default_root_dir=tmpdir, precision=16, strategy=ColossalAIStrategy())
+
+    assert isinstance(trainer.strategy, ColossalAIStrategy)
+    assert isinstance(trainer.strategy.precision_plugin, ColossalAIPrecisionPlugin)
+
+
+@RunIf(colossalai=True)
+def test_colossalai_strategy_with_trainer_by_string(tmpdir):
+    trainer = Trainer(fast_dev_run=True, default_root_dir=tmpdir, precision=16, strategy="colossalai")
+
+    assert isinstance(trainer.strategy, ColossalAIStrategy)
+    assert isinstance(trainer.strategy.precision_plugin, ColossalAIPrecisionPlugin)
+
+
+class ModelParallelBoringModel(BoringModel):
+    def __init__(self):
+        super().__init__()
+        self.layer = None
+
+    def configure_sharded_model(self) -> None:
+        self.layer = torch.nn.Linear(32, 2)
+
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        self.configure_sharded_model()
```

Comment:
If the strategy configures `restore_checkpoint_after_setup = True` (it's a property), then this wouldn't be necessary I believe. The trainer would call configure_sharded_model() and then load the weights after that.

Reply:
Thanks. I've configured it to True now.